File: Easy/PYTHON/DepthFirstSearch.py
class Node:
    def __init__(self, name):
        self.name = name
        self.children = []

    def add_child(self, name):
        self.children.append(Node(name))
        return self

    visited = []

    def depth_first_search(self, visited):
        self.visited.append(self.name)
        for child in self.children:
            if child.name not in self.visited:
                child.depth_first_search(child)

        return self.visited


graph = Node('A')
graph.add_child('B').add_child('C').add_child('D')
graph.children[0].add_child('E').add_child('F')
graph.children[2].add_child('G').add_child('H')
graph.children[0].children[1].add_child('I').add_child('J')
graph.children[2].children[0].add_child('K')


result = graph.depth_first_search(graph)
print(result)

File: Easy/PHP/DepthFirstSearch.php
<?php

namespace Easy\PHP\DepthFirstSearch;

class Node
{
    public $name;
    public $children = array();
    public $visited = array();
 
    public function __construct($name)
    {
        $this->name = $name;
    }

    
 
    public function depthFirstSearch($node)
    {
        array_push($this->visited, $node->name);
        foreach ($node->children as $child) {
            if (in_array($node->name, $this->visited)) {
                $this->depthFirstSearch($child);
            }
        }

        return $this->visited;
    }

    public function addChild($name)
    {
        array_push($this->children, new Node($name));
        return $this;
    }
}

$graph = new Node('A');
$graph->addChild('B')->addChild('C')->addChild('D');
$graph->children[0]->addChild('E')->addChild('F');
$graph->children[2]->addChild('G')->addChild('H');
$graph->children[0]->children[1]->addChild('I')->addChild('J');
$graph->children[2]->children[0]->addChild('K');

$result = $graph->depthFirstSearch($graph);
//var_dump("$result");
// print_r("Graph: ".json_encode($result));
print_r("Graph: ".implode(",", $result));
 